### PR TITLE
Handle multiple peers

### PR DESCRIPTION
**Pull request by Hasan6979**
In order to support meshnet on esp32, we need to be able to add/delete and update multiple peers. The current esp32 wireguard supported only 1 peer.

Another thing that has changed is that the lookup was being done by index, but that is now via encoded public keys. The public key saved in wireguard peer config is a decoded base64 public key.

### CODE REVIEW

File: include/esp_wireguard.h [hunk 1]
```diff
@@ -124,7 +137,7 @@ esp_err_t esp_wireguard_set_default(wireguard_ctx_t *ctx);
 /**
  * @brief Test if the peer is up.
  */
-esp_err_t esp_wireguardif_peer_is_up(wireguard_ctx_t *ctx);
+esp_err_t esp_wireguardif_peer_is_up(wireguard_ctx_t *ctx, wireguard_peer_config_t *peer_config);
```
Comment (matislovas) [inline include/esp_wireguard.h]: Maybe it is enough to pass a pub key, to check if peer is up ..?

File: src/esp_wireguard.c [hunk 2]
```diff
@@ -168,15 +166,15 @@ static esp_err_t esp_wireguard_netif_create(const wireguard_config_t *config)
     wg.listen_port = config->listen_port;
     wg.bind_netif = NULL;
 
-    ESP_LOGI(TAG, "allowed_ip: %s", config->allowed_ip);
+    ESP_LOGI(TAG, "allowed_ip: %s", config->base_ip);
```
Comment (matislovas) [inline src/esp_wireguard.c]: Change the log to `base_ip`

File: CMakeLists.txt [hunk 3]
```diff
@@ -10,9 +10,14 @@ idf_component_register(
          "src/crypto/refc/x25519.c"
          "src/esp_wireguard.c"
          "src/nacl/crypto_scalarmult/curve25519/ref/smult.c"
+         "src/derp.c"
```
Comment (matislovas) [inline CMakeLists.txt]: Maybe it is better to exclude derp from this PR and include it in another with all the DERP stuff ..?

File: examples/demo/main/main.c [hunk 4]
```diff
@@ -49,25 +49,31 @@ static EventGroupHandle_t s_wifi_event_group;
...
-    wg_config.endpoint = CONFIG_WG_PEER_ADDRESS;
-    wg_config.port = CONFIG_WG_PEER_PORT;
-    wg_config.persistent_keepalive = CONFIG_WG_PERSISTENT_KEEP_ALIVE;
+    wg_config.base_ip = CONFIG_WG_LOCAL_IP_ADDRESS;
+    wg_config.net_mask = CONFIG_WG_LOCAL_IP_NETMASK;
+    peer_config.public_key = CONFIG_WG_PEER_PUBLIC_KEY;
+    // if (strcmp(CONFIG_WG_PRESHARED_KEY, "") != 0) {
+    //     wg_config.preshared_key = CONFIG_WG_PRESHARED_KEY;
+    // } else {
+    peer_config.preshared_key = NULL;
```
Comment (matislovas) [inline examples/demo/main/main.c]: Maybe it is worth adding second peer and try to ping it as well ..? Just from a smoke testing perspective :smile:

File: examples/demo/main/main.c [hunk 5]
```diff
@@ -396,26 +408,30 @@ void app_main(void)
...
+    ESP_LOGI(TAG, "Connecting.");
+    err = esp_wireguard_add_peer(&ctx, &peer_config);
+    if (err != ESP_OK) {
+        ESP_LOGE(TAG, "esp_wireguard_connect: %s", esp_err_to_name(err));
+        goto fail;
     }
+    ESP_LOGI(TAG, "Peer is up");
+    vTaskDelay(1000 * 10 / portTICK_PERIOD_MS);
 
+    peer_config.allowed_ip = "10.0.2.3";
```
Comment (olekoliinyk) [inline examples/demo/main/main.c]: I know this is an example, but maybe worth moving to Kconfig.projbuild?

**Comment by olekoliinyk** [comment]
+1 lgtm

**Comment by matislovas** [comment]
+1

**Review by matislovas — APPROVED** [review]

**Pull request merged**